pyomo/contrib/pynumero/interfaces/external_pyomo_model.py

```diff
@@ -190,14 +283,57 @@ def set_input_values(self, input_values):
         for var, val in zip(input_vars, input_values):
             var.set_value(val)
 
+        vector_scc_idx = 0
         for block, inputs in self._scc_list:
             if len(block.vars) == 1:
                 calculate_variable_from_constraint(
                     block.vars[0], block.cons[0]
                 )
             else:
-                with TemporarySubsystemManager(to_fix=inputs):
-                    solver.solve(block)
+                if self._use_cyipopt:
+                    # Transfer variable values into the projected NLP, solve,
+                    # and extract values.
+
+                    nlp = self._vector_scc_nlps[vector_scc_idx]
+                    proj_nlp = self._vector_proj_nlps[vector_scc_idx]
+                    input_coords = self._vector_scc_input_coords[vector_scc_idx]
+                    cyipopt = self._cyipopt_solvers[vector_scc_idx]
+                    _, local_inputs = self._vector_scc_list[vector_scc_idx]
+
+                    primals = nlp.get_primals()
+                    variables = nlp.get_pyomo_variables()
+
+                    # Set values and bounds from inputs to the SCC.
+                    # This works because values have been set in the original
+                    # pyomo model, either by a previous SCC solve, or from the
+                    # "global inputs"
+                    for i, var in zip(input_coords, local_inputs):
+                        # Set primals (inputs) in the original NLP
+                        primals[i] = var.value
+                    # This affects future evaluations in the ProjectedNLP
+                    nlp.set_primals(primals)
+                    x0 = proj_nlp.get_primals()
+                    sol, _ = cyipopt.solve(x0=x0)
+
+                    # Set primals from solution in projected NLP. This updates
+                    # values in the original NLP
+                    proj_nlp.set_primals(sol)
+                    # I really only need to set new primals for the variables in
+                    # the ProjectedNLP. However, I can only get a list of variables
+                    # from the original Pyomo NLP, so here some of the values I'm
+                    # setting are redundant.
+                    new_primals = nlp.get_primals()
+                    assert len(new_primals) == len(variables)
+                    for var, val in zip(variables, new_primals):
+                        var.set_value(val)
+
+                else:
+                    # Use a Pyomo solver to solve this strongly connected
+                    # component.
+                    with TemporarySubsystemManager(to_fix=inputs):
+                        solver.solve(block)
+
+                vector_scc_idx += 1
```

Comment:
I think this is technically correct as written, but it would be clearer if it was done inside the `if self._use_cyipopt` block.

Reply:
I think I like having this outside the `if self._use_cyipopt` tree, as otherwise, in the case where we use a regular Pyomo solver, `vector_scc_idx` will always be zero, even though we may not be at the zeroth "vector-valued" SCC. Having this information may be useful, e.g. for debugging, even if it's not strictly necessary for the algorithm.